includes/wikia/nirvana/WikiaView.class.php

```diff
@@ -114,36 +109,106 @@ protected function buildTemplatePath( $controllerClass, $methodName, $forceRebui
-			$controllerBaseName = $app->getBaseName( $controllerClass );
-			if ($app->isService($controllerClass)) {
-				$controllerClass = $app->getServiceClassName( $controllerBaseName );
-			} else {
-				$controllerClass = $app->getControllerClassName( $controllerBaseName );
+			$controllerClass = $this->normalizeControllerClass( $controllerClass );
+
+			$templateExists = false;
+			$templatePath = '';
+			$templates = $this->getTemplateOptions( $controllerClass, $methodName );
+			$dirName = $this->getTemplateDir( $controllerClass );
+			foreach ( $templates as $templateName ) {
+				$templatePath = $dirName . '/' . $templateName . '.' . $extension;
+
+				if ( file_exists( $templatePath ) ) {
+					$templateExists = true;
+					break;
+				}
 			}
 
-			if( empty( $wgAutoloadClasses[$controllerClass] ) ) {
-				throw new WikiaException( "Invalid controller or service name: {$controllerClass}" );
+			if ( !$templateExists ) {
+				throw new WikiaException( "Template file not found: {$templatePath}" );
 			}
 
-			// First we look for BaseName_MethodName
-			$dirName = dirname( $wgAutoloadClasses[$controllerClass] );
-			$basePath = "{$dirName}/templates/{$controllerBaseName}_{$methodName}";
-			$templatePath = "{$basePath}.$extension";
-			$templateExists = file_exists( $templatePath );
+			$this->setTemplatePath( $templatePath );
+		}
+		wfProfileOut(__METHOD__);
+	}
 
-			// Fall back to ControllerClass_MethodName
-			if( !$templateExists ) {
-				$templatePath = "{$dirName}/templates/{$controllerClass}_{$methodName}.$extension";
-				$templateExists = file_exists( $templatePath );
-			}
+	/**
+	 * For legacy reasons we sometimes get a $controllerClass name that doesn't end with
+	 * 'Controller' or 'Service'.  Normalize to this form.
+	 *
+	 * @param $controllerClass
+	 *
+	 * @return string
+	 *
+	 * @throws WikiaException
+	 */
+	public static function normalizeControllerClass( $controllerClass ) {
+		$app = F::app();
+		$controllerBaseName = $app->getBaseName( $controllerClass );
+		if ( $app->isService( $controllerClass ) ) {
+			$controllerClass = $app->getServiceClassName( $controllerBaseName );
+		} else {
+			$controllerClass = $app->getControllerClassName( $controllerBaseName );
+		}
 
-			if( !$templateExists ) {
-				throw new WikiaException( "Template file not found: {$templatePath}" );
+		if ( empty( $app->wg->AutoloadClasses[$controllerClass] ) ) {
+			throw new WikiaException( "Invalid controller or service name: {$controllerClass}" );
+		}
+
+		return $controllerClass;
+	}
+
+	/**
+	 * Generates a list of possible template names ordered by preference
+	 *
+	 * @param $controllerClass
+	 * @param $methodName
+	 *
+	 * @return array
+	 */
+	public function getTemplateOptions( $controllerClass, $methodName ) {
+		$templates = [];
+
+		// Make sure this method exists, otherwise the call to getMethod crashes PHP
+		// so badly it can't even log that a problem occurred.
+		if ( method_exists( $controllerClass, $methodName ) ) {
```

Comment:
I'd add an request-wide cache here for cases when you render many items separately so this doesn't add a lot of overhead (reflection used to be slow).